sim/core/rppm.go

```diff
@@ -0,0 +1,321 @@
+// It multiplies the actual proc chance by 1 + coefficient
+func (proc *RPPMProc) WithClassMod(coefficient float64, classMask int) *RPPMProc {
+	mod := rppmClassMod{
+		classMask:   classMask,
+		coefficient: coefficient,
+	}
+
+	if proc.char != nil {
+		proc.charCoefficient *= mod.GetCoefficient(proc.char)
+	} else {
+		proc.charMods = append(proc.charMods, mod)
+	}
+
+	return proc
+}
+
+// Attaches a spec mod to the RPPM
+// It multiplies the actual proc chance by 1 + coefficient
+func (proc *RPPMProc) WithSpecMod(coefficient float64, spec proto.Spec) *RPPMProc {
+	mod := rppmSpecMod{
+		spec:        spec,
+		coefficient: coefficient,
+	}
+
+	if proc.char != nil {
+		proc.charCoefficient *= mod.GetCoefficient(proc.char)
+	} else {
+		proc.charMods = append(proc.charMods, mod)
+	}
+
+	return proc
+}
+
+// Set the base coefficient for the RPPM proc chance
+// It's a multiplicator for the RPPM based proc chance
+func (proc *RPPMProc) WithCoefficient(coefficient float64) *RPPMProc {
+	proc.coefficient = coefficient
+	return proc
+}
+
+// Attach an approximate Ilvl scaling to the Mod
+// The proc chance will be multiplied by 1.00936^(ilvlDiff)
+func (proc *RPPMProc) WithApproximateIlvlMod(coefficient float64, baseIlvl int32, realIlvl int32) *RPPMProc {
+	proc.mod = append(proc.mod, rppmApproxIlvlMod{
+		coefficient: coefficient,
+		baseIlvl:    int32(baseIlvl),
+		realIlvl:    realIlvl,
+	})
+
+	return proc
+}
+
+func (proc *RPPMProc) ForCharacter(character *Character) *RPPMProc {
+	proc.char = character
+
+	for _, mod := range proc.charMods {
+		proc.charCoefficient *= mod.GetCoefficient(character)
+	}
+
+	proc.charMods = []rppmCharMod{}
+	return proc
+}
+
+// Create a new RPPM Proc with the given ppm (usually from the ProcsPerMinute record)
+func NewRPPMProc(ppm float64) *RPPMProc {
+	proc := &RPPMProc{
+		ppm:             ppm,
+		coefficient:     1,
+		charCoefficient: 1,
+		lastProc:        -time.Second * 120,
+		lastCheck:       -time.Second * 10,
+		mod:             []rppmMod{},
+	}
+
+	return proc
+}
+
+// Does not change the state of the RPPMProc.
+// Only calculates the proc chance.
+//
+// To actually modify the state correctly call:
+//
+//	Proc(character, sim)
+func (proc *RPPMProc) getProcChance(unit *Unit, sim *Simulation) float64 {
+	basePpm := proc.ppm
+
+	baseCoeff := proc.coefficient
+	for _, mod := range proc.mod {
+		baseCoeff *= mod.GetCoefficient(unit)
+	}
+
+	lastCheck := math.Min(10.0, float64((sim.CurrentTime-proc.lastCheck)/time.Second))
+	lastProc := math.Min(1000.0, float64((sim.CurrentTime-proc.lastProc)/time.Second))
+
+	procCoefficient := math.Max(1, 1+((lastProc/(60/proc.ppm))-1.5)*3) // Bad luck protection
+	baseProcChance := (basePpm * (lastCheck / 60.0)) * baseCoeff
+	return baseProcChance * procCoefficient * proc.charCoefficient
+}
+
+func (proc *RPPMProc) Proc(unit *Unit, sim *Simulation, label string) bool {
+	result := sim.Proc(proc.getProcChance(unit, sim), label)
+	proc.lastCheck = sim.CurrentTime
```

Comment:
Add a comment here clarifying that this models a non-Blizzlike variant of the 5.4.8 system with the first proc bug fixed, with a TODO to revisit once Blizzard makes a decision on how to deal with this interaction.